src/test/java/com/fasterxml/jackson/databind/ser/CanonicalNumberGenerator.java

```diff
@@ -0,0 +1,66 @@
+package com.fasterxml.jackson.databind.ser;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+
+import com.fasterxml.jackson.core.JsonGenerationException;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.util.JsonGeneratorDelegate;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.SerializerProvider;
+
+public class CanonicalNumberGenerator extends JsonGeneratorDelegate {
+
+    /**
+     * TODO The constant should be public or the verify method. Copied from
+     * `jackson-databing` class `NumberSerializer`
+     */
+    protected final static int MAX_BIG_DECIMAL_SCALE = 9999;
+
+    private final ValueToString<BigDecimal> _bigDecimalToString;
+
+    public CanonicalNumberGenerator(JsonGenerator gen, ValueToString<BigDecimal> bigDecimalToString) {
+        super(gen);
+        this._bigDecimalToString = bigDecimalToString;
+    }
+
+    @Override
+    public void writeNumber(double v) throws IOException {
+        BigDecimal wrapper = BigDecimal.valueOf(v);
```

Comment:
This likely changes some values, fwtw (`double` is 2-based, `BigDecimal` 10-based; lossy conversion)?
Or would that only be problematic in reverse direction (I know `0.1` etc are exact in 10-based and transcendental in 2-based; not sure if reverse cases exist).

Reply:
There is a bug in `BigDecimal(double)` but that has been fixed with `BigDecimal.valueOf()`. valueOf() works correctly and always creates a correct BigDecimal, even for numbers that can't be represented with 2-based / IEEE math.

Since my code to canonicalize numbers is string based, we could maybe skip the conversion to `BigDecimal` but the string manipulation was ugly hack until we have a better solution.

The other option would be to use `Number` instead of `BigDecimal` in `ValueToString` and rename it to `NumberToString`. Then we could wrap the primitive in a `Double` instance here and both code paths would look the same plus it would also be possible to process BigInteger if someone wanted to.